pymonetdb/filetransfer.py

```diff
@@ -0,0 +1,621 @@
+"""
+Classes related to file transfer requests as used by COPY INTO ON CLIENT.
+"""
+# This Source Code Form is subject to the terms of the Mozilla Public
+# License, v. 2.0.  If a copy of the MPL was not distributed with this
+# file, You can obtain one at http://mozilla.org/MPL/2.0/.
+#
+# Copyright 1997 - July 2008 CWI, August 2008 - 2016 MonetDB B.V.
+
+
+from abc import ABC, abstractmethod
+import codecs
+from importlib import import_module
+from io import BufferedIOBase, BufferedWriter, TextIOBase, TextIOWrapper
+from pathlib import Path
+from shutil import copyfileobj
+from typing import Optional, Union
+from pymonetdb import mapi
+from pymonetdb.exceptions import OperationalError, ProgrammingError
+
+
+def handle_file_transfer(mapi, cmd: str):
+    if cmd.startswith("r "):
+        parts = cmd[2:].split(' ', 2)
+        if len(parts) == 2:
+            try:
+                n = int(parts[0])
+            except ValueError:
+                pass
+            return handle_upload(mapi, parts[1], True, n)
+    elif cmd.startswith("rb "):
+        return handle_upload(mapi, cmd[2:], False, 0)
+    elif cmd.startswith("w "):
+        return handle_download(mapi, cmd[2:], True)
+    elif cmd.startswith("wb "):
+        return handle_download(mapi, cmd[3:], False)
+    else:
+        pass
+    # we only reach this if decoding the cmd went wrong:
+    mapi._putblock(f"Invalid file transfer command: {cmd!r}")
+
+
+def handle_upload(mapi, filename, text_mode, offset):
```

Comment:
missing type annotations

Reply:
Ok, I've added a whole bunch of additional type annotations.
Not easy because the IO types don't type annotate well.
In particular, for some reason a `BufferedIOBase` is not accepted as an `IO[bytes]`.